### ISSUE TITLE
Plugins could avoid warnings by installing UIApplicationDelegate methods dynamically

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#9984]
### Issue about warning message on console
https://github.com/flutter/engine/pull/8843

Right now iOS apps have the following in their logs:

> You've implemented -[ application:didReceiveRemoteNotification:fetchCompletionHandler:], but you still need to add "remote-notification" to the list of your supported UIBackgroundModes in your Info.plist.

A possible improvement would be to avoid implementing that method directly, and instead only add an implementation using `class_addMethod` if we detect that it's needed for registered plugins. If no plugin implements `application:didReceiveRemoteNotification:fetchCompletionHandler:` then `FlutterAppDelegate` wouldn't implement it either, and thus would avoid the warning.

The same could be done for other `UIApplicationDelegate` methods as well.

/cc @mravn-google @szakarias @jakobr-google @mit-mit @Hixie @goderbauer 

------------------------------------------------------------------------------------------------------------------------

Edited by cyanglaz:

_The original issue mentioned above has been fixed with https://github.com/flutter/engine/pull/8843. However, there is a related but different issue regarding receiving email warning from Apple when uploading builds. I have summarized the discussions and works related to it in below section for easier understanding the bug and where we are at:_

### Issue about warning email from Apple

When uploading a Flutter App to Apple, one might receive an warning Email like below:

```
Dear Developer,We identified one or more issues with a recent delivery for your app, "Flutter Gallery". Your delivery was successful, but you may wish to correct the following issues in your next delivery:Missing Push Notification Entitlement - Your app appears to register with the Apple Push Notification service, but the app signature's entitlements do not include the "aps-environment" entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the "aps-environment" entitlement. Xcode 8 does not automatically copy the aps-environment entitlement from provisioning profiles at build time. This behavior is intentional. To use this entitlement, either enable Push Notifications in the project editor's Capabilities pane, or manually add the entitlement to your entitlements file. For more information, see https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/HandlingRemoteNotifications.html#//apple_ref/doc/uid/TP40008194-CH6-SW1.After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to iTunes Connect.Best regards,The App Store Team
```

Although this warning doesn't block releasing Apps, this warning is annoying and confusing for Flutter Apps that doesn't use push notifications.

### Steps to reproduce

1. You need to have a valid Apple Distribution certs and private keys installed on your computer.
1. Create a new Flutter App. Build it for iOS with `flutter build ios`, or `flutter build ios --local-engine=ios_release` local engine build.
1. With the Apple Distribution account, upload the App to App Store Connect. You can follow the steps here: https://help.apple.com/xcode/mac/current/#/dev442d7f2ca
1. After a successful upload, you will see an email with a message similar to above.

### Workaround

To not receiving this email anymore, you can workaround by enabling publish notification in your App. (You don't have to implement any publish notification features, just simply enable the feature and capability will suppress the warning.)

### Cause

The root cause of this issue is the appearance of [[application:didRegisterForRemoteNotificationsWithDeviceToken:]](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622958-application?language=objc) in the engine code base is the cause of the email warning in https://github.com/flutter/flutter/issues/9984#issuecomment-405380407. 
Either directly implement this method, or dynamically implement this method with a similar solution in https://github.com/flutter/engine/pull/8843 (which only mentioning this method with a @selector syntax) triggers this warning email.

**Comment by mravn-google** [comment]
I assume from the above that it is possible and safe to add a new method after plugins are registered and before the method in question is called? I read [somewhere](http://nshipster.com/method-swizzling/) that method swizzling tricks should be done in the `+load` method which may be too early.

Then there is the issue that people might implement the method in their own `AppDelegate`. Currently they should just call `super` to have the call forwarded to any interested plugins. If we use `class_addMethod`, we need another way of doing this. For instance, we could have a `FlutterAppDelegate` method that does plugin forwarding, but using a different selector like (`forwardToPluginsApplication:didReceiveRemoteNotification:fetchCompletionHandler:`). The method we add should call this one.

**Comment by collinjackson** [comment]
Urban Airship solves this problem with swizzling in `application:didFinishLaunching:` so I think it's safe for us to do so as well. It wouldn't make sense to do it in `[FlutterAppDelegate load]` because that would get called before `[GeneratedPluginRegistrant registerWithRegistry:]` so we wouldn't know what the plugins are yet.

My current thinking is that we should remove the plugin-related `UIApplicationDelegate` implementations from `FlutterAppDelegate`. Instead, let plugins swizzle the `[UIApplication sharedApplication].delegate` object directly in their registerWithRegistrar implementation (possibly providing an API for doing so, so that devs don't have to learn how to swizzle correctly and they can write plugins in Swift someday).

If AppDelegate provides a `didReceiveRemoteNotification:fetchCompletionHandler:` implementation, the call to [super didReceiveRemoteNotification:fetchCompletionHandler:] will do nothing because it's not implemented by FlutterAppDelegate. It'll work regardless of whether they call the super method or not.

**Comment by csga5000** [comment]
Where's this at?

**Comment by eseidelGoogle** [comment]
This bug causes developers to get this warning after uploading to the AppStore:

""" 
Dear Developer,We identified one or more issues with a recent delivery for your app, "Flutter Gallery". Your delivery was successful, but you may wish to correct the following issues in your next delivery:Missing Push Notification Entitlement - Your app appears to register with the Apple Push Notification service, but the app signature's entitlements do not include the "aps-environment" entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the "aps-environment" entitlement. Xcode 8 does not automatically copy the aps-environment entitlement from provisioning profiles at build time. This behavior is intentional. To use this entitlement, either enable Push Notifications in the project editor's Capabilities pane, or manually add the entitlement to your entitlements file. For more information, see https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/HandlingRemoteNotifications.html#//apple_ref/doc/uid/TP40008194-CH6-SW1.After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to iTunes Connect.Best regards,The App Store Team
"""

FYI @mjohnsullivan this might warrant a mention in https://flutter.io/ios-release/  or maybe it's already covered in stack overflow.

**Comment by pulyaevskiy** [comment]
Is there documentation on how to workaround this issue while proposed solution is in progress, it'd be very helpful.

**Comment by csga5000** [comment]
@pulyaevskiy 
I'm pretty sure the "solution" is ignore the warnings.  They don't technically break anything.  Apple will still approve your app, as long as you're not actually using push notifications.

**Comment by jamesderlin** [comment]
Instead of swizzling, would it be sufficient if `FlutterAppDelegate` instead used `respondsToSelector:` and `forwardingTargetForSelector:` and made them conditionally forward to the `FlutterPluginAppLifeCycleDelegate`?

**Comment by darrenangle** [comment]
> This bug causes developers to get this warning after uploading to the AppStore:
> 
> """ 
> Dear Developer,We identified one or more issues with a recent delivery for your app, "Flutter Gallery". Your delivery was successful, but you may wish to correct the following issues in your next delivery:Missing Push Notification Entitlement - Your app appears to register with the Apple Push Notification service, but the app signature's entitlements do not include the "aps-environment" entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the "aps-environment" entitlement. Xcode 8 does not automatically copy the aps-environment entitlement from provisioning profiles at build time. This behavior is intentional. To use this entitlement, either enable Push Notifications in the project editor's Capabilities pane, or manually add the entitlement to your entitlements file. For more information, see https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/HandlingRemoteNotifications.html#//apple_ref/doc/uid/TP40008194-CH6-SW1.After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to iTunes Connect.Best regards,The App Store Team
> """
> 
> FYI @mjohnsullivan this might warrant a mention in https://flutter.io/ios-release/ or maybe it's already covered in stack overflow.


Still getting this warning on every upload.

**Comment by gurugeek** [comment]
getting this warning too ! Ignoring is not the solution really.

**Comment by KevinManders** [comment]
We get the same thing, 
Cordova app - build with adobe phonegap build
Plugin:
phonegap-plugin-push   	npm	~2.2.3	2.2.3	android,ios,winphone

On ad-hoc build, no problem. But on the release build:
Same mail as stated above "missing entitlement", but i get one more error/problem.
The push wont register and gives the error:
`{"line":62,"column":50,"sourceURL":"http://localhost:8080/plugins/phonegap-plugin-push/www/push.js"}`.
Tried:
- Other versions of the plugin;
- Cordova update;
- New provisioning cert.

**Comment by PandaGeek1024** [comment]
Im getting the same warning, would be good if we can remove this warning.

**Comment by nohli** [comment]
As a workaround for getting rid of the Apple warning mails after uploading (regarding missing push notifications entitlement) you can just activate push notifications under certificates in Xcode.
The iOS user won't even be asked for permissions for push notifications, nor will it show up in settings.

**Comment by cyanglaz** [comment]
https://github.com/flutter/engine/pull/8843 is merged. Now the issue should be resolved. I am closing the issue. Please feel free to reopen if any related issues are found.

**Comment by PerLycke** [comment]
@eseidel As you pointed out the issue with the warning after uploading to AppStore, was this suppose to fix this? I still get the warning even after this was merged. Now this is closed, and the issue pointing directly to the warning issue is also closed. Everything closed but hopefully not forgotten as my CEO still believes something is wrong with our app (he gets the warning email too) :D

**Comment by eseidelGoogle** [comment]
My understanding is this issue is not fixed, since I also still get an email warning every time we publish Flutter's own example apps. :/

@collinjackson, @amirh do you know?

**Comment by eseidelGoogle** [comment]
Actually, it's very possible this is fixed on master, but not on `stable` or `beta` and it's possible that we're just currently releasing our example apps from one of those (and that @PerLycke your company is also releasing from a stable/beta branch and thus don't have the fix yet)?

**Comment by PerLycke** [comment]
Posted the above after I released from dev earlier today to see if it was fixed, but still got the warning email. So not fixed in dev where this is merged @eseidelGoogle

**Comment by cyanglaz** [comment]
This fix should start at flutter v1.7.0

**Comment by PerLycke** [comment]
> This fix should start at flutter v1.7.0 @PerLycke are you seeing the warning in the original post in the console as well?

I got the warning email using:

> Doctor summary (to see all details, run flutter doctor -v):
> [✓] Flutter (Channel dev, v1.7.9, on Mac OS X 10.14.5 18F132, locale en-SE)
>  
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
> [✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
> [✓] iOS tools - develop for iOS devices
> [✓] Chrome - develop for the web
> [✓] Android Studio (version 3.4)
> [✓] IntelliJ IDEA Community Edition (version 2019.1.3)
> [✓] Connected device (2 available)
> 
> • No issues found!

**Comment by PerLycke** [comment]
@cyanglaz Didn't see your edit before I posted. No, I don't see that in the console. Just this annoying email from Apple. If there is anything else I can try to help please let me know

**Comment by cyanglaz** [comment]
@PerLycke Hmm. it seems like my fix only fixed the log. Apple would still send email even if the delegate methods are implemented dynamically.
One thing you can do to workaround it for now until we fix it is to enable the push notification capability for your iOS app.
You can follow here https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app?language=objc
specifically under `Enable the Remote Notifications Capability`
Remember the check both remote notification and background fetch. Please let me know if this helps.

**Comment by collinjackson** [comment]
I put some ideas for a potential solution into flutter/engine#9488

@PerLycke would you be willing to try patching your engine with this change and submit your app and let us know if you still get the warning?

https://github.com/flutter/flutter/wiki/Compiling-the-engine

**Comment by PerLycke** [comment]
@cyanglaz The workaround you proposed didn't work and I still got the warning email from Apple. I guess the last resort is to enable push notifications to avoid the warning email.

@collinjackson I'm sorry but I'm struggling building for ios using a custom built engine, and I don't know how to get around the error. So I can't confirm if your fix works or not :/ Getting the following error:

> Failed to build iOS app
> Error output from Xcode build:
> ↳
>     ** BUILD FAILED **
> 
> 
> Xcode's output:
> ↳
>     ld: warning: Could not find auto-linked framework 'Flutter'
>     Undefined symbols for architecture armv7:
>       "_FlutterMethodNotImplemented", referenced from:
>           -[FLTDeviceInfoPlugin handleMethodCall:result:] in DeviceInfoPlugin.o
>       "_OBJC_CLASS_$_FlutterMethodChannel", referenced from:
>           objc-class-ref in DeviceInfoPlugin.o
>     ld: symbol(s) not found for architecture armv7
>     clang: error: linker command failed with exit code 1 (use -v to see invocation)
>     /Users/per/android/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_secure_storage-3.2.1+1/ios/Classes/FlutterSecureStoragePlugin.m:87:49: warning: values of type 'OSStatus' should not be used as format arguments; add an explicit cast to 'int' instead
>     [-Wformat]
>                 NSLog(@"SecItemUpdate status = %d", status);
>                                                ~~   ^~~~~~
>                                                %d   (int)
>     /Users/per/android/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_secure_storage-3.2.1+1/ios/Classes/FlutterSecureStoragePlugin.m:95:46: warning: values of type 'OSStatus' should not be used as format arguments; add an explicit cast to 'int' instead
>     [-Wformat]
>                 NSLog(@"SecItemAdd status = %d", status);
>                                             ~~   ^~~~~~
>                                             %d   (int)
>     2 warnings generated.
> 
> Encountered error while building for device.

**Comment by elmariocarlos** [comment]
Just submitted and ap and got this; and I do have the profile and everything setup correctly, so not sure whats is going on.

**Comment by collinjackson** [comment]
There are some helpful notes in firebase-ios-sdk explaining how the warning email is triggered and how to get around it.

https://github.com/firebase/firebase-ios-sdk/blob/7eb1747e4b9be195ea9908b9da28b1d0f818afe3/GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m#L98

```
 *  We have to opt out of referencing APNS related App Delegate methods directly to prevent
 *  an Apple review warning email about missing Push Notification Entitlement
 *  (like here: https://github.com/firebase/firebase-ios-sdk/issues/2807). From our experience, the
 *  warning is triggered when any of the symbols is present in the application sent to review, even
 *  if the code is never executed. Because GULAppDelegateSwizzler may be used by applications that
 *  are not using APNS we have to refer to the methods indirectly using selector constructed from
 *  string.
 *
 *  NOTE: None of the methods is proxied unless it is explicitly requested by calling the method
 *  +[GULAppDelegateSwizzler proxyOriginalDelegateIncludingAPNSMethods]
```

I've updated flutter/engine#9488 with an approach based on this (waiting on @RedBrogdon to try submitting an app for me).

Relatedly, it seems that Firebase's swizzling got broken recently, see the discussion in https://github.com/flutter/flutter/issues/35267 and https://github.com/flutter/plugins/pull/1826

**Comment by cyanglaz** [comment]
I am currently working on some other stuff and won't be able to actively work on this recently. I am going to un-assign myself from this issue.
cc @collinjackson

**Comment by midhunarmid** [comment]
Any update on this?
I recently uploaded an application and received the above-mentioned warning as email.

There are no other issues, but the warning email itself is very annoying and giving a feel like something went wrong in the application.

**Comment by javipacheco** [comment]
Same here. I have received the same email this morning. Any update about this?

**Comment by nohli** [comment]
> As a workaround for getting rid of the Apple warning mails after uploading (regarding missing push notifications entitlement) you can just activate push notifications under certificates in Xcode.
> The iOS user won't even be asked for permissions for push notifications, nor will it show up in settings.

**Comment by elmariocarlos** [comment]
I can confirm with @nohli this seems to be the only way to get around this; unfortunatelly we still depend on having the macOS device for preparation purposes. This really needs to get looked at eventually, as it takes away options of using tools like codemagic to avoid the expenses associated with adquiring 2 computers to get the job done.

**Comment by rodruiz** [comment]
You can add a file "**Runner.entitlements**" under ios/Runner/Runner.entitlements
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>aps-environment</key>
	<string>development</string>
</dict>
</plist>
```

You have to enable "Push Notifications" under Capabilities. Xcode will generate the Runner.entitlements file for you.

![xcode-fix-aps-env](https://user-images.githubusercontent.com/10534494/64391244-e9a31c80-cffc-11e9-8851-34156852d93d.png)

**Comment by kaushikb1996** [comment]
> You can add a file "**Runner.entitlements**" under ios/Runner/Runner.entitlements
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
> <plist version="1.0">
> <dict>
> 	<key>aps-environment</key>
> 	<string>development</string>
> </dict>
> </plist>
> ```
> 
> You have to enable "Push Notifications" under Capabilities. Xcode will generate the "**Runner.entitlements**" file for you.
> 
> ![xcode-fix-aps-env](https://user-images.githubusercontent.com/10534494/64391244-e9a31c80-cffc-11e9-8851-34156852d93d.png)

My configuration is exactly as you have shown but still getting the warming email.

**Runner.entitlements**

`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>aps-environment</key>
	<string>development</string>
</dict>
</plist>
`

![Screenshot 2019-09-18 at 8 06 56 PM](https://user-images.githubusercontent.com/5248406/65158423-ef165300-da4f-11e9-8549-d91758b7e34c.png)

![Screenshot 2019-09-18 at 8 08 24 PM](https://user-images.githubusercontent.com/5248406/65158500-18cf7a00-da50-11e9-804f-a3a06e0985b4.png)

**Comment by hagen00** [comment]
I too can't get rid of the warning email. I would consider this to be more than just an annoyance because it means it's now very difficult to get users to approve the app. With this error, TestFlight is rejecting the app. Even though the actual release may not be rejected, how do people approve if not via Test Flight?

**Comment by nohli** [comment]
@hagen00 in the comment above yours, the workaround is explained pretty well, did you try it?

**Comment by hagen00** [comment]
@nohli I tried setting the two push notification permissions in Xcode, but didn't add the Runner.entitlements file. The app is live now, but next time I try get the app onto Testflight, I'll try add the Runner.entitlements file as well.

**Comment by borolepratik** [comment]
I can confirm that @rodruiz's solution worked for the Flutter stable channel! I do not get that warning message anymore when submitting a revision to App Store Connect.

**Comment by Skyost** [comment]
Then the file named by @rodruiz (`Runner.entitlements`) should be generated alongside other Xcode files by `flutter create`, isn't it ?

**Comment by bottee** [comment]
I think this issue and it's proposed workaround is a severe security risk and I can't understand why it is proposed to enable the capability for push notifications and to create an entitlement for a functionality even when it is not used. 
If any flutter app has to enable this and there happens to be a malicious pub who sends malicious push notifications to all apps which are using this pub you will have a great problem. 
Please correct me if I'm wrong

**Comment by mithunmqb** [comment]
So the safest solution is still to ignore the warning ?

**Comment by sjghose** [comment]
> So the safest solution is still to ignore the warning ?

Appears like that...looking for clarifications as well.

**Comment by sgehrman** [comment]
same issue.  3 years and still no solution?

**Comment by rr0055256** [comment]
Any update on the solution ? Faced this issue, when the build was pushed to Appstore.

**Comment by 77Y** [comment]
i still receive this "ITMS-90078: Missing Push Notification Entitlement" email, so any solutions?

**Comment by rezz2020** [comment]
I have this problem, any update?

**Comment by chinmaygarde** [comment]
This seems to have been fixed.

**Comment by PerLycke** [comment]
@chinmaygarde When you say it's fixed, Is it fixed in master? I've just uploaded to App Store and still got this email: https://github.com/flutter/flutter/issues/9984#issuecomment-405380407

flutter doctor -v:

```
[✓] Flutter (Channel beta, 1.21.0-9.2.pre, on Mac OS X 10.15.6 19G2021, locale en-SE)
    • Flutter version 1.21.0-9.2.pre at /Users/per/dev/flutter
    • Framework revision 81a45ec2e5 (4 days ago), 2020-08-27 14:14:33 -0700
    • Engine revision 20a9531835
    • Dart version 2.10.0 (build 2.10.0-7.3.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/per/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
```

**Comment by nguyenvangiangbn** [comment]
> This seems to have been fixed.

I check in stable channel still have problem

**Comment by JamesBewley** [comment]
@chinmaygarde This is still an issue in stable today

```
Flutter 1.20.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 216dee60c0 (8 days ago) • 2020-09-01 12:24:47 -0700
Engine • revision d1bc06f032
Tools • Dart 2.9.2
```

**Comment by florisdipt** [comment]
Neither these solutions work for me, any way to override all the notifications handlers with a dummy class

**Comment by TugayYaldiz** [comment]
We have this issue as well and it worries us. Is there any fix planned for this issue in the near future? Is there any risk of apps being rejected?

**Comment by nohli** [comment]
> You can add a file "**Runner.entitlements**" under ios/Runner/Runner.entitlements
> 
> ```xml
> 
> <?xml version="1.0" encoding="UTF-8"?>
> 
> <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
> 
> <plist version="1.0">
> 
> <dict>
> 
> 	<key>aps-environment</key>
> 
> 	<string>development</string>
> 
> </dict>
> 
> </plist>
> 
> ```
> 
> 
> 
> You have to enable "Push Notifications" under Capabilities. Xcode will generate the "**Runner.entitlements**" file for you.
> 
> 
> 
> ![xcode-fix-aps-env](https://user-images.githubusercontent.com/10534494/64391244-e9a31c80-cffc-11e9-8851-34156852d93d.png)
> 
> 

This is the workaround.

**Comment by marcqtan** [comment]
Any fix yet? I'm still encountering this issue in flutter 1.22.2 build. I don't want to use the workaround since I would need to generate a new provisioning profile for that.

**Comment by nohli** [comment]
> Any fix yet? I'm still encountering this issue in flutter 1.22.2 build. I don't want to use the workaround since I would need to generate a new provisioning profile for that.

@marcqtan actually, if you just enable it in Xcode (like in the screenshot above), there will be no provisioning profile under https://developer.apple.com/account/resources/profiles/list.

**Comment by zenkog** [comment]
For this workaround, my question is will the user be asked to consent to Pushed Notification? Since we don't use it we don't want to send a conflicting message to the user.

**Comment by luis901101** [comment]
@zenkog 
No, the user will not be asked about anything related to push notifications... in fact the Notifications permission access will not be visible in the app settings from the OS settings.

**Comment by mauriziobarucco** [comment]
Sorry, but does the @nohli 's workaround enable the push notifications capabilities on the app? If so, I think is a security threat

**Comment by bradyt** [comment]
Grumble.

**Comment by bradyt** [comment]
We now have both this issue and https://github.com/flutter/flutter/issues/14182 closed. @chinmaygarde has not replied to requests for clarification on why they closed this issue.

The workaround comes after it was closed. But it seems to contradict itself, and now I feel the need to research why "enabling push notifications" does not really "enable push notifications" but actually just stops the email warnings about push notification.

In my opinion, the other issue should be reopened, as it is more aptly named. This issue is titled with a particular approach to solving it, if I understand correctly. The issue developers are facing is described more directly at https://github.com/flutter/flutter/issues/14182.

I am curious why @Hixie removed this from Milestones. I do see https://github.com/flutter/flutter/issues/61969 was spontaneously closed before then. I'm guessing because this issue's title is too specific in how this issue should be solved.

**Comment by bradyt** [comment]
EDIT: the other issue was reopened, please disregard all but the second paragraph. You can even disregard that, to some extent. 😀

**Comment by jmagman** [comment]
@bradyt I got ahead of myself, I think https://github.com/flutter/flutter/issues/14182 is properly a duplicate of this one.

I'm reopening this until we can get more clarification from either the plugin or engine team about why this was closed.  https://github.com/flutter/engine/pull/6259 and https://github.com/flutter/engine/pull/9488 were closed, not merged.

The last time this seems touched was https://github.com/flutter/engine/pull/8843

**Comment by bradyt** [comment]
I think I found a small change in the workaround that made a little more sense to me. In xcode, I added an entitlement, for example push notifications, then immediately removed it. It left behind this "empty entitlements file",

```diff
new file   ios/Runner/Runner.entitlements
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict/>
+</plist>
```

As well as a few lines like,

```diff
modified   ios/Runner.xcodeproj/project.pbxproj
@@ -30,6 +30,7 @@
 /* End PBXCopyFilesBuildPhase section */
 
 /* Begin PBXFileReference section */
+		006286AE25EDB32300B9A629 /* Runner.entitlements */ = {isa = PBXFileReference; lastKnownFileType = text.plist.entitlements; path = Runner.entitlements; sourceTree = "<group>"; };
 		1498D2321E8E86230040F4C2 /* GeneratedPluginRegistrant.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = GeneratedPluginRegistrant.h; sourceTree = "<group>"; };
 		1498D2331E8E89220040F4C2 /* GeneratedPluginRegistrant.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; path = GeneratedPluginRegistrant.m; sourceTree = "<group>"; };
 		3B3967151E833CAA004F5970 /* AppFrameworkInfo.plist */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.xml; name = AppFrameworkInfo.plist; path = Flutter/AppFrameworkInfo.plist; sourceTree = "<group>"; };
```

I gave it a try pushing to TestFlight, and this time I did not receive an email warning from Apple about push notifications. This makes a little more sense to me, rather than entitling a push notification to avoid a warning, just adding an "empty" entitlement file to avoid a warning.

**Comment by bradyt** [comment]
Can anyone clarify why the workaround makes sense? If my app does not use push notifications, why is the fix for this to add entitlement for push notifications? Can the clarification include a quote from official documentation?

**Comment by bradyt** [comment]
Just to add clarification on why this is difficult to ignore, since I think some have suggested to ignore.

The subject line of the email is,

> App Store Connect: Your app "<app name>" (Apple ID: <id> Version: <version> Build: <build>) has one or more issues

So we must be alarmed at email, check contents, verify that there is only one issue, and that it is the issue about push notifications.

Presumably it would be tricky to write an email filter for this.

**Comment by dpedrinha** [comment]
Developing for Apple is already a pain. They are always complaining about something. It's a matter of time until they complain about setting push and not using it or not activating it at their console.

**Comment by eli-front** [comment]
Apple is just trying to ignore the fact that so many people don't develop natively anymore. More support for Flutter = happy devs = more cool apps.

**Comment by mdridley** [comment]
The other issue I see with the workaround, is that you cannot use push notification entitlements for wildcard certs... It seems right now that you will get a provisioning profile missing entitlement error right now on wildcard attempts to build.

**Comment by cyanglaz** [comment]
I have spent some time to explore different solutions. It seems like the appearance of [[application:didRegisterForRemoteNotificationsWithDeviceToken:]](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622958-application?language=objc) in the engine code base is the cause of the email warning in https://github.com/flutter/flutter/issues/9984#issuecomment-405380407. 

I have updated the issue description to make this Email warning issue clearer for easier future reference.

**Comment by benwinding** [comment]
Over 4 years later now... is there any plan to fix this? 😟 

It's not a great developer experience when apple says you have `1 or more issues` on every single release...

**Comment by Renesanse** [comment]
Stuck on this problem too...

**Comment by gaaclarke** [comment]
I've implemented a fix based on the information provided above: https://github.com/flutter/engine/pull/29996

Is there anyone in the position and willing to test this change quickly?  It will take me a minute to setup the conditions to reproduce it.

**Comment by gaaclarke** [comment]
This issue is closed as a result of https://github.com/flutter/engine/pull/29996.  Since this is an old and much upvoted issue I just want to remind everyone that issues are closed when the fix lands on the master channel.  You can still run into this problem on stable until the PR is merged into the stable channel.  It's progress can be tracked in its roll commit into the Flutter repository: [**Roll commit TBD**].  That commit will be tagged with the versions where the fix is present.

I'll also add that this issue involves a black box from Apple so we can't automate tests and Apple could change the algorithm they use for triggering the notification at anytime.  Testing it manually is awkward, but assuming the comments above and what Firebase implemented is correct, this should be fixed now as the remote notification symbols no longer show up in Flutter library.

Thanks!  Let me know if any issues arise.

**Comment by jmagman** [comment]
Reopening, https://github.com/flutter/engine/pull/29996 was reverted in https://github.com/flutter/engine/pull/30330 due to b/210427076

> Unfortunately it causes the [1P plugin] to not receive any APNS registration token and therefore not receive any push notifications. I'd guess it ends up breaking external push messaging plugins as well.

> Aha there is a bad interaction here between the flutter delegate swizzling and the firebase SDK swizzling (not the firebase plugin but the native SDK itself). If I comment out firebase's swizzling for didRegister then the chime plugin does receive the callback.

The swizzling worked on its own, but unfortunately Firebase and Flutter swizzling the same methods stomped on each other.  Need to investigate something similar that doesn't interfere with Firebase's attempts to avoid the same warnings.

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by gaaclarke** [comment]
I talked @jklint-g and the swizzling that Firebase does when initialized isn't related to getting callbacks for Flutter plugins.  It happens regardless of Flutter.

https://github.com/flutter/engine/pull/29996 added the register methods for FlutterPluginAppLifeCycleDelegate at runtime and did not add them to the FlutterAppDelegate, instead relying on the forwarding target, which broke initialization of Firebase.  Firebase needs those present when its init is called.  If we also add those register methods at runtime to the FlutterAppDelegate everything should work.  The order of operations should be correct since I was editing the runtime at `+load` time, but Firebase's swizzling is happening when init is called for the Firebase API.

**Comment by phamconganh** [comment]
So now we still have to add Push Notification even though we don't use it?

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by jmagman** [comment]
> So now we still have to add Push Notification even though we don't use it?

No, you do not need the entitlement unless your app uses Push. This warning is safe to ignore.

> Although this warning doesn't block releasing Apps, this warning is annoying and confusing for Flutter Apps that doesn't use push notifications.

**Comment by mchlsndrn** [comment]
> No, you do not need the entitlement unless your app uses Push. This warning is safe to ignore.
That much was clear from the beginning.
Judging from the reactions on some comments, it looks like lots of people find this warning annoying and not a great user experience.
Is this issue going to be closed with the recommendation "please ignore the warning" ?

**Comment by jmagman** [comment]
> Is this issue going to be closed with the recommendation "please ignore the warning" ?

No, the intention is to fix it, which is why the issue is still open.  There have been several attempts to fix it but all have caused different, worse issues and have been reverted.

**Comment by marianhlavac** [comment]
Can this be fixed? It's ridiculous.

**Comment by gsouf** [comment]
@marianhlavac  Saying "it's ridiculous" wont help to fix the issue and is disrespectful to people putting efforts on working on flutter

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by larry-cariq** [comment]
Just ran into this issue with Flutter 3.3.2. Looking forward to a resolution soon.

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by SunlightBro** [comment]
Apple moved to restrict AdHoc to businesses with 100+ employees.
Smaller flutter developers now have to work inside their clients Apple-Business-Manager.

Now every time I upload a new app version to TestFlight,
not just me but the **client** gets this warning email about "Missing Push Notification Entitlement".

So now it's no longer an inconvenience that I can just ignore,
but I have to explain this issue to clients who thinks it's unprofessional and I can't even provide them an ETA when this will be resolved.

**Comment by haoyaxin** [comment]
您好，您的邮件我已经收到，我会尽快给您回复，谢谢！
                                                       ——郝雅欣

**Comment by eseidel** [comment]
Just wanted to report that this is still an issue, the Flutter counter app, when pushed to the App Store produces this email:
```
ITMS-90078: Missing Push Notification Entitlement - Your app appears to register with the Apple Push Notification service, but the app signature's entitlements do not include the 'aps-environment' entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the 'aps-environment' entitlement. Xcode does not automatically copy the aps-environment entitlement from provisioning profiles at build time. This behavior is intentional. To use this entitlement, either enable Push Notifications in the project editor's Capabilities pane, or manually add the entitlement to your entitlements file. For more information, see https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/HandlingRemoteNotifications.html#//apple_ref/doc/uid/TP40008194-CH6-SW1.

After you’ve corrected the issues, you can upload a new binary to App Store Connect.
```

**Comment by mit-mit** [comment]
cc @leighajarett

**Comment by gaaclarke** [comment]
It's been a couple of years so people might have missed some history on this issue.  In november 2021 I "fixed" this issue in https://github.com/flutter/engine/pull/29996.

The reason that these warnings happens is because Apple's submit process looks for the presence of certain symbols in the binary, and warns about entitlements if it sees them.  They are there for push notifications.  So, if your app isn't using push notifications, the warnings are irrelevant.

That PR removed those methods from the Flutter binary, adding them at runtime.  The problem is that there are plugins that are built on the assumption that those methods exist so that they can swizzle in their own behavior.  Breaking clients who are swizzling the engine is not defined as a Breaking Change.  However in this case, I'm not entirely sure why since I was absent, it was decided that it would be preferable to revert the fix to accommodate the the plugin that was swizzling and any hypothetical plugin that may be doing the same thing.

So, in summary it's a conflict of `Apple warning if certain symbols are present in your binary` and `plugins that rely on those symbols being present to be used unorthodoxly`.

Possible solutions:
1) Try to find some way to guarantee that our swizzling will happen before any other swizzling so the methods are present when they try to swizzle.
1) Find some way to more clearly break plugins who rely on those symbols being present.  (I'm guessing there was already an error that they just ignore when they tried to swizzle something that didn't exist.)
1) Relitigate the revert of the fix (and reach out to migrate plugins we know about or add in a hack to the engine to accommodate them somehow)

**Comment by b055man** [comment]
@pattobrien has a good point though: like you said, although the "warnings are irrelevant", it still does look really unprofessional if you need to explain stakeholders that this warning received from Apple, for every build you push, isn't really a problem. Plus it becomes annoying if you frequently push updates to TestFlight and this email is sent to everyone, for every. single. build.

**Comment by pattobrien** [comment]
@gaaclarke thanks for the insight into your prior fix attempt.

> ...it was decided that it would be preferable to revert the fix to accommodate the the plugin that was swizzling and any hypothetical plugin that may be doing the same thing.

I'm sure there's more to this than what I'm about to suggest, but perhaps we should consider evaluating the effects of this issue on 100% of iOS Flutter apps, versus the effects of the breaking change on a small subset of push notification apps. I myself have not shipped a single app (of 6 that I've worked on) that requires push notifications, since many use cases can simply be covered by one-off or scheduled notifications via a package like `local_notifications`. The `firebase_messaging` package is in the top 80 most liked packages on pub.dev, so breaking changes obviously need to be considered - but at what expense?

**Comment by GorinMaxim** [comment]
any news?

**Comment by TitouanBourveauEXFO** [comment]
any update ?

**Comment by jgwinner** [comment]
Adding onto this thread - we've recently run into the same problem.

I agree with @pattobrien and @b055man - our stakeholders look at this askance. Heck, I only found out because I followed up with our developer to add the entitlement, not realizing this was a long standing annoyance. 

It's crazy this has been open for 6+ years.

**Comment by jmagman** [comment]
Let's keep conversation focused on how to fix the issue.  We know it's a problem, it's an open issue, and you can see several engineers have tackled it in the past.  "I agree" just adds noise.

**Comment by Waskar** [comment]
I solved the problem by following the steps in the first two comments of the following link: https://stackoverflow.com/questions/32251123/missing-push-notification-entitlement

**Comment by wickyou23** [comment]
Hello guys, is there any solution for this issue? I don't use remote notifications, so I don't want to create an aps-environment.

**Comment by mackayn** [comment]
A bit janky to ask people to enable push notifications for dev only, I'll stick with the warning.

**Comment by hasanmhallak** [comment]
disabling automatic signing and manually sign the app should resolve this issue.

**Comment by PerLycke** [comment]
> disabling automatic signing and manually sign the app should resolve this issue.

Do you mean Xcode automatic signing? How would that solve it, technically? You are just proposing a solution without saying how exactly it would resolve the issue.

Also, I don't and I've never used automatic signing and I always had this issue. Also, this issue dates back 7 years and I'm sure many with me aren't using automatic signing and still have this issue. Which directly contradicts what you are proposing.

**Comment by hasanmhallak** [comment]
@PerLycke Honestly I'm just confused as you are right now, I tried with an old project and it worked. but now after your comment, I just tried it with our current project and sadly it didn't work. I guess I'll edit the other comment.

**Comment by eseidel** [comment]
Replying to Aaron above: https://github.com/flutter/flutter/issues/9984#issuecomment-1664282440

There might also be a 4th way, which Apple uses itself for issues like this which is adding a runtime "quirk" to only add the symbol for certain plugins.  e.g. We should be able to find all Plugins this would break by scanning pub (running `nm` or `strings` on the ios binaries searching for this symbol).  And if so building a list of plugins that need this symbol, and then only registering it from the engine (at runtime) when those binaries/bundleids are present within the application.

Here are some examples from "quirks" in WebKit (there are a ton more in AppKit, depending on what app AppKit is loaded into to maintain old compatibility behavior, but AppKit isn't open source):
https://github.com/search?q=repo%3AWebKit%2FWebKit+quirk+language%3AObjective-C%2B%2B&type=code&l=Objective-C%2B%2B
(It looks like they don't have the list of sites or apps they quirk for checked into GitHub, but they do have the quirking infrastructure.)

But probably the easier solution is to just revert the revert and announce a breaking change.  Right now this bug results in millions of mails from Apple to Flutter devs every time they push to the store, which as others have commented, "makes flutter look bad".

**Comment by kjxbyz** [comment]
I have this problem, any update?

flutter: 3.24.3

```
ITMS-90078: Missing potentially required entitlement - Your app, or a library that’s included in your app, uses Apple Push Notification service (APNs) registration APIs, but the APS Environment Entitlement isn’t included in the app signature’s entitlements. If your app uses APNs, make sure the App ID is enabled for push notifications in Certificates, Identifiers & Profiles, sign the app with a distribution provisioning profile that includes the APS Environment Entitlement, and upload a new build to ensure that push notifications function as intended. For details, visit: https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns.
```

**Comment by felixkrautschuk** [comment]
I am getting the same warning as @kjxbyz 
we are not using Push notifications at all in our app.

These are the dependencies our app uses:
  ```
cupertino_icons: ^1.0.8
  package_info_plus: ^8.1.0
  http: ^1.2.2
  provider: ^6.1.2
  font_awesome_flutter: ^10.7.0
  intl: any
  shared_preferences: ^2.3.2
  icons_launcher: ^3.0.0
```

**Comment by abjainri** [comment]
And this issue is still present and it is Dec 2024. I got the mail from App Store Connect while I pushed my build to test flight only.

**Comment by alexwastl** [comment]
Can confirm that we get this message from Apple as well in Dec 2024 however not using push notifications or related packages in our flutter app.

**Comment by stevo-knievo** [comment]
Same here. I just pushed my build to TestFlight and received the email from App Store Connect. My Flutter app does not use push notifications or related packages.

Here’s the email I received:
```
Hello,

We noticed one or more issues with a recent delivery for the following app:

- FooBar
- Version 1.0.0
- Build 1

This is for your information only and doesn't require any action:

ITMS-90078: Missing potentially required entitlement - Your app, or a library that’s included in your app, uses Apple Push Notification service (APNs) registration APIs, but the APS Environment Entitlement isn’t included in the app signature’s entitlements. If your app uses APNs, make sure the App ID is enabled for push notifications in Certificates, Identifiers & Profiles, sign the app with a distribution provisioning profile that includes the APS Environment Entitlement, and upload a new build to ensure that push notifications function as intended. For details, visit: https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns.

```
I have double-checked everything:

1. Verified the Apple Developer Portal settings (Certificates, Identifiers & Profiles) to ensure that the Push Notifications capability is correctly disabled.
2. Regenerated the provisioning profile to reflect the updated App ID configuration.
3. Confirmed that no references to push notifications exist in my app's code or dependencies.

Despite all this, the warning persists. Any insights would be greatly appreciated.

**Comment by abjainri** [comment]
> Same here. I just pushed my build to TestFlight and received the email from App Store Connect. My Flutter (version 3.27.1) app does not use push notifications or related packages.
> 
> Here’s the email I received:
> 
> ```
> Hello,
> 
> We noticed one or more issues with a recent delivery for the following app:
> 
> - FooBar
> - Version 1.0.0
> - Build 1
> 
> This is for your information only and doesn't require any action:
> 
> ITMS-90078: Missing potentially required entitlement - Your app, or a library that’s included in your app, uses Apple Push Notification service (APNs) registration APIs, but the APS Environment Entitlement isn’t included in the app signature’s entitlements. If your app uses APNs, make sure the App ID is enabled for push notifications in Certificates, Identifiers & Profiles, sign the app with a distribution provisioning profile that includes the APS Environment Entitlement, and upload a new build to ensure that push notifications function as intended. For details, visit: https://developer.apple.com/documentation/usernotifications/registering-your-app-with-apns.
> ```
> 
> I have double-checked everything:
> 
> 1. Verified the Apple Developer Portal settings (Certificates, Identifiers & Profiles) to ensure that the Push Notifications capability is correctly disabled.
> 2. Regenerated the provisioning profile to reflect the updated App ID configuration.
> 3. Confirmed that no references to push notifications exist in my app's code or dependencies.
> 
> Despite all this, the warning persists. Any insights would be greatly appreciated.

Just add Push notifications capability in Xcode for your project even if you are not using it.

**Comment by LionWY** [comment]
only add the notification， 8 years。。。

**Comment by androidseb** [comment]
Hello, I just noticed that I'm no longer receiving these email notifications from Apple (I had gotten used to receiving a blast of 3 notifications instead of 2 with each new build uploaded).

I received one of these notifications last on April 6 for an app build built with Flutter 2.27. The next build after that was submitted on April 29, still with Flutter 2.27 and triggered the other usual email notifications about the build being processed, but not the email notification about missing push entitlement.

It seems somewhere between April 6 and April 29, Apple has changed their warning emails strategy?

With the recent builds I've been submitting over the last couple of weeks with Flutter 3.32, I didn't receive these missing entitlement notifications either, can anyone else confirm this is no longer an issue for them either?

**Comment by kjxbyz** [comment]
> Hello, I just noticed that I'm no longer receiving these email notifications from Apple (I had gotten used to receiving a blast of 3 notifications instead of 2 with each new build uploaded).
> 
> I received one of these notifications last on April 6 for an app build built with Flutter 2.27. The next build after that was submitted on April 29, still with Flutter 2.27 and triggered the other usual email notifications about the build being processed, but not the email notification about missing push entitlement.
> 
> It seems somewhere between April 6 and April 29, Apple has changed their warning emails strategy?
> 
> With the recent builds I've been submitting over the last couple of weeks with Flutter 3.32, I didn't receive these missing entitlement notifications either, can anyone else confirm this is no longer an issue for them either?

Apple should have solved this problem
